interfaces/builtin/dbus.go

```diff
@@ -0,0 +1,418 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/release"
+)
+
+const dbusPermanentSlotAppArmor = `
+# Description: Allow owning a name on DBus public bus
+
+#include <abstractions/###DBUS_ABSTRACTION###>
+
+# register on DBus
+dbus (send)
+    bus=###DBUS_BUS###
+    path=/org/freedesktop/DBus
+    interface=org.freedesktop.DBus
+    member="{Request,Release}Name"
+    peer=(name=org.freedesktop.DBus, label=unconfined),
+
+dbus (send)
+    bus=###DBUS_BUS###
+    path=/org/freedesktop/DBus
+    interface=org.freedesktop.DBus
+    member="GetConnectionUnix{ProcessID,User}"
+    peer=(name=org.freedesktop.DBus, label=unconfined),
+
+# bind to a well-known DBus name: ###DBUS_NAME###
+dbus (bind)
+    bus=###DBUS_BUS###
+    name=###DBUS_NAME###,
+
+# For KDE applications, also support alternation since they use org.kde.foo-PID
+# as their 'well-known' name. snapd does not allow ###DBUS_NAME### to end with
+# '-[0-9]+', so this is ok.
+dbus (bind)
+    bus=###DBUS_BUS###
+    name=###DBUS_NAME###-[1-9]{,[0-9]}{,[0-9]}{,[0-9]}{,[0-9]}{,[0-9]},
+
+# Allow us to talk to dbus-daemon
+dbus (receive)
+    bus=###DBUS_BUS###
+    path=###DBUS_PATH###
+    peer=(name=org.freedesktop.DBus, label=unconfined),
+dbus (send)
+    bus=###DBUS_BUS###
+    path=###DBUS_PATH###
+    interface=org.freedesktop.DBus.Properties
+    peer=(name=org.freedesktop.DBus, label=unconfined),
+`
+
+const dbusPermanentSlotAppArmorClassic = `
+# allow unconfined clients to introspect us on classic
+dbus (receive)
+    bus=###DBUS_BUS###
+    interface=org.freedesktop.DBus.Introspectable
+    member=Introspect
+    peer=(label=unconfined),
+
+# allow unconfined clients talk to ###DBUS_NAME### on classic
+dbus (receive, send)
+    bus=###DBUS_BUS###
+    interface=###DBUS_INTERFACE###
+    peer=(label=unconfined),
+
+# allow unconfined to all interfaces via ###DBUS_PATH### (eg,
+# org.freedesktop.*, org.gtk.Application, etc) to allow integrating in classic
+# environment.
+dbus (receive, send)
+    bus=###DBUS_BUS###
+    path=###DBUS_PATH###
+    peer=(label=unconfined),
+`
+
+const dbusPermanentSlotSecComp = `
+# Description: Allow owning a name on DBus public bus
+getsockname
+recvmsg
+sendmsg
+sendto
+`
+
+const dbusPermanentSlotDBus = `
+<policy user="root">
+    <allow own="###DBUS_NAME###"/>
+    <allow send_destination="###DBUS_NAME###"/>
+</policy>
+<policy context="default">
+    <allow send_destination="###DBUS_NAME###"/>
+</policy>
+`
```

Comment:
dbusPermanentSlotDBus wasn't in the original PR but it is needed for 'bus=system'. @niemeyer and @tyhicks: the question then becomes, should non-root be allowed to talk to it? As implemented (in a separate commit for easy reverting), the answer is 'yes' because snapd should allow a non-root snap command should be able to connect to a root daemon within the same command and it is up to the snap to perform any uid checks. However, because of the dbusPermanentSlotAppArmorClassic policy for on classic (and the lack of policykit in the core snap) this means that unconfined non-root can talk to the snap-- this needs discussion because this interface does alone not grant extra access to the system, b) this interface in combination with privileged ones (eg, network-control) requires a user to connect the interfaces and c) unconfined non-root can aa_change_profile() to something that would allow the access. 

'b' is somewhat mediated in the initial store upload since the base declaration requires that to use this interface you need to claim the name and receive a snap declaration. However, the tools aren't considering combinations of interfaces and it would be a shame if a system, root-running slot implementation service has say, network-control and exposes a DBus API but doesn't perform uid checks so non-root users could escalate privileges to change networking via the poorly implemented snap. Granted, there is nothing saying this poorly written snap couldn't expose a named unix socket in SNAP_DATA world read/write and to the same thing without this interface, so this is probably one of those things where this poorly written gets poor reviews and/or bug report and we exercise store policies to compel the developer to fix it (or revoke the snap).

@niemeyer and @tyhicks, should non-root be able to contact a dbus system service via this interface?

(sorry for the second email on this, I made a change and the comment was hidden so I deleted the first and added this one)